src/flog.rs

```diff
@@ -79,6 +79,8 @@ pub mod categories {
 
         (warning_path, "warning-path", "Warnings about unusable paths for config/history (on by default)", true);
 
+        (deprecated_test, "deprecated-test", "Warning about using test's zero- or one-argument modes (`test -d $foo`), which will be changed in future.");
```

Comment:
Fine by me.

Given what I know I'm fine with enabling this warning now already
(and I'd probably not change the exit code but I'm not gonna fight)
In the light of that I'd also rephrase "will change" to "may change"

Reply:
To be clear: As this PR stands, *by default*, it will not do anything.

If the warning is enabled, it will show the warning, the exit code will remain.

If the feature flag is enabled, the exit code will change.

The end goal is to have the behavior that I consider best, which is that `test -n` and `test -z` are fine with no additional arguments, anything else is an error.

The plan is to stage this carefully:

- People can, once this is merged, enable the warning. Once they've fixed their code, they can set the feature flag
- The next step is to enable the warning, but keep the flag off by default, so the exit codes stay
- The step after that is to enable the warning and the flag, but allow disabling them
- The step after that is to make the flag read-only and remove the warning - now we have the behavior we would want from `test`

That means we have *a step* where there's a warning but no change in the exit code (by default), but it's only a step. I do not consider "warn about bad behavior but keep it" to be a good final outcome, I would like to go further and end at "okay behavior".

But I also believe that giving people some time where they *know* the change is coming without already throwing warnings at them - which is, as we have discussed, itself a breaking change for many uses, even if it is easy to revert. That's why I would like to have that first step.

Reply:
I think everything that can be said has been said already. You go ahead.
I don't fully agree but my expertise in this area is quite limited.

Reply:
> The plan is to stage this carefully:
> 
> People can, once this is merged, enable the warning. Once they've fixed their code, they can set the feature flag

yeah you're right that announcing things in release notes before we change anything is the proper way to do it.
Then plugin authors can react to it. Hard to tell if it matters in this case but it's good practice